### PR TITLE
Added optional Select input binding

### PR DESCRIPTION
**Pull request by minettiandrea**
First of all thank you very much for the great product. I'm a very happy udash user! And using it extensivly on my open source project https://github.com/Insubric/box

For my projects I'm using quite a lot the native HTML5 form validation, an in general with udash it works smoothly, I just come across a minor limitation on the select element, using `required` on the select element with the current implementation does not trigger any validation and we have no way to put a `null` option, but there are use cases where this is needed.
For reference see MDN select required documentation https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#attr-required.

In order to overcome this limitation my proposition is to add a `optional` method on `io.udash.bindings.inputs.Select` with that signature:
```
def optional[T](
    selectedItem: Property[Option[T]], options: ReadableSeqProperty[T], labelNoValue:Modifier
  )(label: T => Modifier, selectModifiers: Modifier*): InputBinding[Select]
```
basically is the same of the `apply` just the selectedItem is optional and we need a label when no value is selected.

The usage is pretty straightforward.

Thanks

Andrea

### CODE REVIEW

**Comment by ddworak** [comment]
Thank you for your kind words, it's always great to see Udash successfully used outside of our org :)

Regarding the PR, I'm not yet convinced that it improves upon the current API so much - I think I may be missing some use case. Let's assume:
```scala
val fruitProperty = Property[Option[Fruit]](None)
```

Then the current `Select` could be used like:
```scala
Select(
  fruitProperty, Seq(Apple, Orange, Banana).map(_.option).prepended(None).toSeqProperty
)(maybeFruit => maybeFruit.mapOr("Label no value", _.name))
```
and the new, utility one:
```scala
Select.optional(
  fruitProperty, Seq(Apple, Orange, Banana).toSeqProperty, labelNoValue = "Label no value"
)(_.name)
```

Sure, we get to avoid boxing in `.option` and prepending explicitly, but if the renderer for labels is any more complex, you'll have to call it separately in each arg list. Maybe if you show your example, we can find a nice interface to cover it.

**Comment by minettiandrea** [comment]
Hi @ddworak ,
thank you for replying

my example it's no more complex from what you wrote, the need of adding the `optional` method was for the rendered output, with the current implementation you will have something like:
```html
<select>
   <option value="0">Label no value</option>
   <option value="1">Apple</option>
   ...
</select>
```
but you won't be able to use HTML5 required since "Label no value" has `0` in the value attribute

with my implementation you'll get:
```html
<select>
   <option value="">Label no value</option>
   <option value="0">Apple</option>
   ...
</select>
```
so if "Label no value" is selected and the element is required the validation fails as expected.

But I agree that we could just change the underlying implementation.
Here (L21:SelectBinding.scala) :
```scala
 val el = option(value := idx.toString, label(opt)).render
```
if `opt` is None we can set the value as empty string instead the index, but I don't know if that could break things somewhere else, of course the `onChange` method should be changed accordingly.

If you like that idea I can update the PR with the implementation

File: core/.js/src/test/scala/io/udash/bindings/inputs/SelectTest.scala [hunk 1]
```diff
@@ -125,6 +125,25 @@ class SelectTest extends UdashFrontendTest {
...
       r2.value should be("1")
     }
+
+    "handle optional case" in {
+      val options = Seq(Some("A"), None, Some("B"))
+      val p = Property[Option[Option[String]]](Some(None))
+
+      val select = Select.optional(p, options.toSeqProperty,StringFrag("empty"))(x => StringFrag(x.getOrElse(""))).render
+
+      select.childElementCount should be(4) // empty value should be included
```
Comment (ddworak) [inline core/.js/src/test/scala/io/udash/bindings/inputs/SelectTest.scala:135]: I think you can even go as far as to test with `.outerHTML` here.

File: core/.js/src/main/scala/io/udash/bindings/inputs/Select.scala [hunk 2]
```diff
@@ -25,13 +25,37 @@ object Select {
...
+   * @param label Provides element's label.
+   * @param selectModifiers Additional Modifiers for the select tag, don't use modifiers on value, onchange and selected attributes.
+   * @return Binding with `select` element, which can be used as Scalatags modifier.
+   */
+  def optional[T](
+    selectedItem: Property[Option[T]], options: ReadableSeqProperty[T], labelNoValue:Modifier
+  )(label: T => Modifier, selectModifiers: Modifier*): InputBinding[Select] = {
+    new SelectBinding(options, label, Some(labelNoValue), selectModifiers)(
+      opt => selectedItem.transform(_.contains(opt)),
+      opts => if (!opts.exists(x => selectedItem.get.contains(x))) selectedItem.set(None),
```
[the remark this answers is not in the document]
Comment (ddworak) [reply core/.js/src/main/scala/io/udash/bindings/inputs/Select.scala:66]: `selectedItem.transform(_.contains(opt))` in your original PR was okay (and probably even more idiomatic), I meant the second function. `opts.exists` traverses all `opts` even when `selectedItem` is empty

**Review by ddworak — COMMENTED** [review]
Thank you for these improvements. I've left 2 minor comments and I think we'll be ready to ship after that.

**Comment by minettiandrea** [comment]
Done, when you think it could be available on maven central?

**Review by ddworak — APPROVED** [review]
LGTM! Thank you for nice contribution :) I'll push the M17 tag right after merge, so it'll be available as soon as the central processes the artifact, usually a couple hours.
